### PR TITLE
Change vacations source from 1C to email

### PR DESCRIPTION
**Pull request by sponkrashin**
Issue #390

### CODE REVIEW

File: server/Arcadia.Assistant.InboxEmail.Abstractions/Email.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+            this.UniqueId = uniqueId;
+            this.Text = text;
+            this.Attachments = attachments;
+        }
+
+        public uint UniqueId { get; }
+
+        public string Text { get; }
+
+        public IEnumerable<Stream> Attachments { get; }
```
Comment (santee) [inline server/Arcadia.Assistant.InboxEmail.Abstractions/Email.cs]: is it a message? Stream is not something serializable, it's hard to manage its lifecycle. byte arrays?

Comment (santee) [reply server/Arcadia.Assistant.InboxEmail.Abstractions/Email.cs]: especially since its MemStream :D

Comment (sponkrashin) [reply server/Arcadia.Assistant.InboxEmail.Abstractions/Email.cs]: Fixed.

File: server/Arcadia.Assistant.CSP/Vacations/ArcadiaVacationRegistry.cs [hunk 2]
```diff
@@ -78,8 +87,35 @@ protected override void OnReceive(object message)
...
+
+            await Task.WhenAll(employeesTask, vacationsInfoTask);
+
+            if (vacationsInfoTask.Result is VacationsEmailLoader.GetVacationsInfo.Error errorResult)
+            {
+                throw new Exception("Failed to load employee vacations info", errorResult.Exception);
+            }
+
+            var employees = employeesTask.Result;
+            var vacationsInfo = ((VacationsEmailLoader.GetVacationsInfo.Success)vacationsInfoTask.Result).EmployeeVacations;
```
Comment (santee) [inline server/Arcadia.Assistant.CSP/Vacations/ArcadiaVacationRegistry.cs]: not really like the cast here. It won't throw compiler error in (i know) unlikely event of another added `GetVacationsInfo.Response` subtype. The safe (and readable) way would be to have a switch instead. However, result will be the same, an exception thrown and actor recreated, so i'll leave it to you :)

File: server/Arcadia.Assistant.CSP/Vacations/EmployeesQueryExecutor.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+
+    using Microsoft.EntityFrameworkCore;
+
+    using Arcadia.Assistant.CSP.Model;
+
+    public class EmployeesQueryExecutor
```
Comment (santee) [inline server/Arcadia.Assistant.CSP/Vacations/EmployeesQueryExecutor.cs]: why is this class needed?

File: server/Arcadia.Assistant.CSP/Vacations/VacationsEmailLoader.cs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        private void AfterInitialState(object message)
+        {
+            switch (message)
+            {
+                case GetVacationsInfo _:
+                    var result = new GetVacationsInfo.Success(this.vacationsInfo);
+                    this.Sender.Tell(result);
+                    break;
+
+                case InboxEmailsEventBus msg:
```
Comment (santee) [inline server/Arcadia.Assistant.CSP/Vacations/VacationsEmailLoader.cs]: -> EmailsRecievedEventBus?

File: server/Arcadia.Assistant.CSP/Vacations/VacationsEmailLoader.cs [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+                $"Received vacations email from {lastActualEmail.Sender} with subject {lastActualEmail.Subject}"
+            );
+
+            var vacationsAttachment = Encoding.UTF8.GetString(lastActualEmail.Attachments.First());
+            this.vacationsInfo = this.ParseVacations(vacationsAttachment);
+        }
+
+        private List<EmployeeVacationRecord> ParseVacations(string vacationsAttachment)
+        {
+            var lines = vacationsAttachment.Split(new[] { "\r\n" }, StringSplitOptions.None);
```
Comment (santee) [inline server/Arcadia.Assistant.CSP/Vacations/VacationsEmailLoader.cs]: -> `new[] {"\r\n", "\n" }` to support weird environment scenarios

File: server/Arcadia.Assistant.CSP/Vacations/EmployeesQueryExecutor.cs (hunk 3 above)
Comment (sponkrashin) [reply server/Arcadia.Assistant.CSP/Vacations/EmployeesQueryExecutor.cs]: To not have loading of employees logic in vacations registry itself.
There are 3 parts:
- employees executor which knows how to get all employees by some conditions
- vacations email loader that periodically pushes new vacations information
- vacations registry that combines them together

File: server/Arcadia.Assistant.CSP/Vacations/VacationsEmailLoader.cs (hunk 5 above)
Comment (sponkrashin) [reply server/Arcadia.Assistant.CSP/Vacations/VacationsEmailLoader.cs]: Done.

File: server/Arcadia.Assistant.CSP/Vacations/VacationsEmailLoader.cs (hunk 4 above)
Comment (sponkrashin) [reply server/Arcadia.Assistant.CSP/Vacations/VacationsEmailLoader.cs]: Done.

File: server/Arcadia.Assistant.CSP/Vacations/ArcadiaVacationRegistry.cs (hunk 2 above)
Comment (sponkrashin) [reply server/Arcadia.Assistant.CSP/Vacations/ArcadiaVacationRegistry.cs]: Done.

**Pull request merged**